### PR TITLE
New Branch/Tags Not Available and Git Version Not Updated On First Request

### PR DESCRIPTION
**Pull request by rlynch**
This addresses issue #439 along with some other issues.  This PR was originally addressed just to fix a bug I found where the first request to spring cloud server after a git update results in the prior version number being returned along with the new config.  However, during further testing, I realized that the fix was only partial.  Here is a list of issues it resolved...

- SCCS would return an incorrect version on the first request after the following use cases.
  - An update to an existing branch on the Git server.
  - When retrieving a tag the version would be null
  - When a hardReset was done due to a merge error.
- SCCS would return a NoSuchLabelException when requesting a tag or branch that was created after the initial clone. #439

I have also added some easy setup code to create tests which operate off a local and remote repository which allows for more robust testing of "real world" git scenarios.  I have also added many new unit tests to cover these use cases.

### CODE REVIEW

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JGitEnvironmentRepository.java [hunk 1]
```diff
@@ -171,6 +171,7 @@ private Ref refresh(String application, String label) {
 			Ref ref = checkout(git, label);
 			if (shouldPull(git, ref)) {
 				pull(git, label, ref);
+				ref = git.getRepository().getRef(ref.getName());
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JGitEnvironmentRepository.java]: Please put a comment here as to why this is needed.

**Comment by spencergibb** [comment]
Awesome test BTW

**Comment by rlynch** [comment]
@spencergibb - requested comment is done.  Please review.

**Comment by rlynch** [comment]
@ryanjbaxter and @spencergibb 
After doing some more testing with this fix I discovered some uncovered use cases where this fix didn't return the correct version. Specifically, the system would return the prior version (with correct config) on the first request after the local repo was pulled.  

While looking into this I also discovered why new branches/tags were not showing up if they were created after the initial clone as described in #439. The reason was twofold.
1. Since checkout was being called prior to the pull, any new branches or tags didn't exist locally so it would throw a NoSuchLabelException.
2. The JGit pull doesn't offer the ability to add a --tags so even if we did the pull prior to the checkout it still wouldn't see new tags.  
The fix was to first do a fetch prior to the checkout (unless shouldPull is false) so we can see the new tags/branches.  We also no longer do a pull, but instead do a separate fetch(so we can add the --tags option) and merge.  A pull is basically just a fetch and merge so this is fine.

Fixing both issues were fairly intertwined so I decided to submit this as one PR.  

As an added bonus I have added some good test setup code to make testing git calls with a local repo that replicates with a remote repo we can make changes to.  This makes the unit test more "real".  Several new unit tests were added to cover server-side git changes.

**Comment by Krustie101** [comment]
@rlynch 
In your last comment, you mention you do a separate fetch allowing you to add the "--tag" flag but I do not see this in your code (the TagOpts.FETCH_TAGS option is not set on the FetchCommand). As a consequence, a moved tag will not be fetched from the remote repo.

**Comment by rlynch** [comment]
Wow, it was included in my brain. I will add it along with a test case. Thanks for pointing out my miss!

**Comment by rlynch** [comment]
@Krustie101 - TagOpts.FETCH_ALL added to fetch and unit test extended to cover this case.  Thanks again for finding this.

**Comment by Krustie101** [comment]
@rlynch 
Thank you for putting in the effort to improve the config server. Much appreciated!

**Comment by rlynch** [comment]
@spencergibb @ryanjbaxter - Guys, any comment(s) on this?

**Comment by spencergibb** [comment]
@rlynch I haven't had a chance to review the new changes. It's in my queue.

**Pull request merged**